File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.java
class Solution {
    public int characterReplacement(String s, int k) 
    {
        int i = 0;
        int j = 0;
        int n = s.length();
        int[] freq = new int[26];
        int longest = 0;
        int max_freq = 0;
        while(j<n){
            char ch = s.charAt(j);
            freq[ch-'A']++;
            max_freq = Math.max(max_freq,freq[ch-'A']);
            while((j-i+1)-max_freq > k) {
                freq[s.charAt(i)-'A']--;
                i++;
            }
            longest = Math.max(longest, j-i+1); 
            j++;
        }
        return longest;
    }
}

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        freq = {}
        i = 0
        j = 0
        length = len(s)
        max_length = 0
        while j < length:
            if s[j] not in freq:
                freq[s[j]] = 0
            freq[s[j]] += 1
            if(freq[s[j]] > 1):
                index = i
                for index in range(i,j):
                    freq[s[index]] -= 1
                    i += 1
                    if s[index] == s[j]:
                        break
            max_length = max(max_length,j-i+1)
            j += 1
        return max_length
        